contributors/design-proposals/network/service-topology-key.md

```diff
@@ -0,0 +1,187 @@
+    // "weight" to the sum if the endpoints matches the corresponding topology key; the
+    // endpoints(s) with the highest sum are the most preferred.
+    // The service routing probability is calculated by the sum. For example, a service 
+    // has two endpoints: A and B. If endpoint A has sum 100 and endpoint B has sum 50,
+    // then endpoint A has 2/3 probability while endpoint B has 1/3 probability for the service.
+    // +optional
+    Preferred []WeightedTopologyTerm
+}
+
+// Defines a topolgoy level.
+type TopologyTerm struct {
+    // A topology key, associated with the corresponding weight.
+    // topologyKey is the key for the node label that the system uses to denote such a topology domain. There are some built-in topology keys, e.g. kubernetes.io/hostname,
+    // failure-domain.beta.kubernetes.io/zone,
+    // failure-domain.beta.kubernetes.io/region,
+    // The built-in topology keys can be good examples and we recommend users switch to a similar mode for portability, but it's NOT enforced.     
+    // Users can define whatever topolgoy key they like since topology is arbitrary.
+    // Required.
+    TopologyKey string `json:"topologyKey" protobuf:"bytes,1,opt,name=topologyKey"`
+}
+
+// The weights of all of the matched WeightedTopologyTerm fields are added to find the most preferred service endpoints.
+type WeightedTopologyTerm struct {
+    // weight associated with matching the corresponding topologyKey,
+    // in the range 1-100. Default Weight value is 1 instead of 0!
+    Weight int32 `json:"weight" protobuf:"varint,1,opt,name=weight"`
+    // A topology key, associated with the corresponding weight.
+    // topologyKey is the key for the node label that the system uses to denote such a topology domain. There are some built-in topology keys, e.g. kubernetes.io/hostname,
+    // failure-domain.beta.kubernetes.io/zone,
+    // failure-domain.beta.kubernetes.io/region,
+    // The built-in topology keys can be good examples and we recommend users switch to a similar mode for portability, but it's NOT enforced.     
+    // Users can define whatever topolgoy key they like since topology is arbitrary.
+    // Required.
+    TopologyKey string `json:"topologyKey" protobuf:"bytes,2,opt,name=topologyKey"`
+}
+```
+
+An example of service that uses topologyKey:
+
+```yaml
+apiVersion: v1
+kind: Service
+metadata:
+  name: with-service-topology
+spec:
+  required:
+    - topologyKey: failure-domain.beta.kubernetes.io/zone
+    - topologyKey: failure-domain.beta.kubernetes.io/region
+  preferred:
+    - weight: 100
+      topologyKey: kubernetes.io/hostname
+```
+
+In our example, both "same zone" AND "same region" are required while "same host" is not required but preferred.
+
+#### Endpoints API changes
+
+Although `NodeName` was already added to `EndpointAddress`, we want `Endpoints` to carry more node's topology informations so that allowing more topology levels other than hostname. 
```

Comment:
I donot know much about this. But I wonder whether we can get topology information by nodename.

Reply:
topologyKey `kubernetes.io/hostname` will behaves similar as the current "NodeName". But, the different is topologyKey is arbitrary, and we may migrate the special `NodeName` to topologyKey in the future.

Reply:
We can't get topology informations from nodename as many as we want.

Reply:
I mean to get `NodeLabels ` by `NodeName`. In other word, just put the work by endpoint controller into kube-proxy.

Reply:
>  In other word, just put the work by endpoint controller into kube-proxy.

I let endpoint controller parse node labels because we should need a global node information table to do that. For example, kube-proxy running in host A and endpoint running in host B, what topology keys that the endpoint carries? We won't want kube-proxy watch ALL nodes, right?

Reply:
Seems this is the controller's job.  If put this logic in kube-proxy, then apiserver may not handle so much request pressure.

Good job.

Reply:
> just put the work by endpoint controller into kube-proxy.

That would mean you couldn't use this topology feature with endpoint addresses that are not hosted on a node, which doesn't seem ideal

> Seems this is the controller's job. If put this logic in kube-proxy, then apiserver may not handle so much request pressure.

speaking of performance impact, denormalizing *all* node labels into all endpoint addresses:
* puts a lot more data into endpoints
* makes a node label change have enormous write multiplication

we could mitigate that by only including node labels referenced in the service spec's topology constraints

regardless, the endpoints controller will constantly be checking if node update events are changing labels or not (since all nodes update every ~5 seconds). we'll want to update the large scale test to measure the impact of that

Reply:
> we could mitigate that by only including node labels referenced in the service spec's topology constraints

Yes, it should work.